### PR TITLE
Enable SMP on GR712RC

### PR DESCRIPTION
**Pull request by lukileczo**
## Description

This PR:
- extracts common TLB interface in HAL and adapts `ia32` port to work with the changes
- enables SMP on GR712RC
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: hal/sparcv8leon3/gaisler/gr712rc/_init.S [hunk 1]
```diff
@@ -220,3 +227,71 @@ set_pdir3:
...
+	/* Enable interrupts and wait to be scheduled */
+	mov %psr, %g1
+	andn %g1, (PSR_PIL), %g1
+	wr %g1, %psr
+	nop
+	nop
+	nop
+
+	/* GR712RC errata 1.7.8 */
+	set 0xfffffff0, %g1
```
Comment (agkaminski) [inline hal/sparcv8leon3/gaisler/gr712rc/_init.S:291]: Maybe we should do this before enabling interrupts? I guess there's a chance we might never execute this instruction

**Review by agkaminski — COMMENTED** [review]
One small doubt on my part, otherwise LGTM. I like the TLB code refactor very much

File: hal/sparcv8leon3/gaisler/gr712rc/_init.S (hunk 1 above)
Comment (lukileczo) [reply hal/sparcv8leon3/gaisler/gr712rc/_init.S:291]: It doesn't really matter - this is only needed if we execute the `wr %g0, %asr19` (which basically is a `wfi` instruction).
From manual:
> the first instruction or data (whichever is first) access to the bus after leaving power-down might be performed incorrectly
> the load instruction after the write to %asr19 provides immunity against a data cache failure due to
the error described by this errata

Basically the first access after power-down must not be cached.

Comment (agkaminski) [reply hal/sparcv8leon3/gaisler/gr712rc/_init.S:291]: Ok, now I understand, thanks

**Review by agkaminski — APPROVED** [review]
LGTM

**Pull request merged**